src/main/java/gregtech/common/covers/filter/BaseFilterContainer.java

```diff
@@ -0,0 +1,189 @@
+            this.onFilterInstanceChange.run();
+        }
+    }
+
+    public void onFilterInstanceChange() {
+        dirtyNotifiable.markAsDirty();
+    }
+
+    public void setOnFilterInstanceChange(@Nullable Runnable onFilterInstanceChange) {
+        this.onFilterInstanceChange = onFilterInstanceChange;
+    }
+
+    public boolean showGlobalTransferLimitSlider() {
+        return getMaxTransferSize() > 0 &&
+                (isBlacklistFilter() || !hasFilter() || currentFilter.showGlobalTransferLimitSlider());
+    }
+
+    public int getTransferLimit(int slotIndex) {
+        if (isBlacklistFilter() || currentFilter == null) {
+            return getTransferSize();
+        }
+        return currentFilter.getTransferLimit(slotIndex, getTransferSize());
+    }
+
+    public boolean test(R toTest) {
+        return !hasFilter() || getFilter().test(toTest);
+    }
+
+    public MatchResult<R> match(R toMatch) {
+        if (!hasFilter())
+            return MatchResult.create(true, toMatch, -1);
+
+        return getFilter().match(toMatch);
+    }
+
+    public int getTransferLimit(R stack) {
+        if (isBlacklistFilter() || currentFilter == null) {
+            return getTransferSize();
+        }
+        return currentFilter.getTransferLimit(stack, getTransferSize());
+    }
+
+    /**
+     * Called when the filter slot has changed to a different filter or has been removed
+     * 
+     * @param notify if true, call {@code onFilterInstanceChange()}
+     */
+    protected abstract void onFilterSlotChange(boolean notify);
+
+    public int getTransferSize() {
+        if (!showGlobalTransferLimitSlider()) {
+            return getMaxTransferSize();
+        }
+        return this.transferSize;
+    }
+
+    public void setTransferSize(int transferSize) {
+        this.transferSize = MathHelper.clamp(transferSize, 1, getMaxTransferSize());
+        onFilterInstanceChange();
+    }
+
+    @Deprecated
+    public void adjustTransferStackSize(int amount) {
+        setTransferSize(transferSize + amount);
+    }
+
+    public void setBlacklistFilter(boolean blacklistFilter) {
+        if (hasFilter()) getFilter().setBlacklistFilter(blacklistFilter);
+        onFilterInstanceChange();
+    }
+
+    public boolean isBlacklistFilter() {
+        return hasFilter() && getFilter().isBlacklistFilter();
+    }
+
+    /** Uses Cleanroom MUI */
+    public abstract IWidget initUI(ModularPanel main, GuiSyncManager manager);
+
+    public void writeInitialSyncData(PacketBuffer packetBuffer) {
+        packetBuffer.writeItemStack(this.filterInventory.getStackInSlot(0));
+    }
+
+    public ItemStack readFilterStack(PacketBuffer buffer) {
+        var stack = ItemStack.EMPTY;
+        try {
+            stack = buffer.readItemStack();
+        } catch (IOException ignore) {}
+        this.filterInventory.setStackInSlot(0, stack);
+        return stack;
+    }
+
+    public abstract void readInitialSyncData(@NotNull PacketBuffer packetBuffer);
```

Comment:
This should at least have an implementation for reading the `itemStack` sent in `writeInitialSyncData`. It doesn't make much sense to write the filter by default, but not read the filter by default.